In [1]:
#data load
from os.path import join
import pandas as pd
import numpy  as np

# 주석 해제 후 진행하시면 됩니다.
data = pd.read_csv(join('dacon5', 'train.csv'),error_bad_lines=False)
test = pd.read_csv(join('dacon5', 'test.csv'),error_bad_lines=False)

In [2]:
data.head()

id  rho  650_src  660_src  670_src  680_src  690_src  700_src  710_src  \
0   0   25  0.37950  0.42993  0.52076  0.57166  0.67818  0.75476  0.83580   
1   1   10  0.00000  0.00000  0.01813  0.00000  0.00000  0.01974  0.00321   
2   2   25  0.00000  0.03289  0.02416  0.03610  0.05843  0.09015  0.14944   
3   3   10  0.27503  0.31281  0.32898  0.41041  0.46587  0.52769  0.64369   
4   4   15  1.01521  1.00872  0.98930  0.98874  1.01773  1.01632  1.00009   

   720_src  ...       940_dst       950_dst       960_dst       970_dst  \
0  0.93623  ...           NaN  0.000000e+00           NaN  1.067504e-18   
1  0.00000  ...  1.343132e-08  6.112685e-09  2.130547e-09           NaN   
2  0.18578  ...  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
3  0.73562  ...  2.245998e-10  1.299511e-10  7.782625e-11           NaN   
4  0.98217  ...  1.457955e-13  8.769053e-14           NaN  1.330237e-13   

        980_dst       990_dst    hhb  hbo2     ca    na  
0  5.998949e-18  4.378513e-17   5.59  4.32   8.92  4.29  
1  9.710091e-09           NaN   0.00  2.83   7.25  4.64  
2  1.329725e-18           NaN  10.64  3.00   8.40  5.16  
3  4.088921e-10           NaN   5.67  4.01   5.05  4.35  
4           NaN           NaN  11.97  4.41  10.78  2.42  

[5 rows x 76 columns]

In [5]:
pd.isna(data).sum(),pd.isna(test).sum()

(id            0
 rho           0
 650_src       0
 660_src       0
 670_src       0
            ... 
 990_dst    1987
 hhb           0
 hbo2          0
 ca            0
 na            0
 Length: 76, dtype: int64,
 id            0
 rho           0
 650_src       0
 660_src       0
 670_src       0
            ... 
 950_dst    1949
 960_dst    2020
 970_dst    1976
 980_dst    2011
 990_dst    1970
 Length: 72, dtype: int64)

In [10]:
pd.isna(data).sum().sum(),pd.isna(test).sum().sum()

(70168, 70000)

In [6]:
data.shape, test.shape

((10000, 76), (10000, 72))

In [16]:
data.describe()

rho       650_src       660_src       670_src       680_src  \
count  10000.000000  10000.000000  10000.000000  10000.000000  10000.000000   
mean      17.568000      0.180212      0.203529      0.229804      0.259158   
std        5.595847      0.272859      0.288661      0.306340      0.324849   
min       10.000000      0.000000      0.000000      0.000000      0.000000   
25%       15.000000      0.007318      0.009520      0.011270      0.013728   
50%       20.000000      0.052025      0.064600      0.078565      0.095355   
75%       25.000000      0.221117      0.276952      0.339075      0.417478   
max       25.000000      1.019990      1.019970      1.019930      1.019880   

            690_src       700_src       710_src       720_src       730_src  \
count  10000.000000  10000.000000  10000.000000  10000.000000  10000.000000   
mean       0.289975      0.322244      0.354101      0.384492      0.412503   
std        0.342362      0.357700      0.370451      0.379054      0.382319   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.016350      0.019510      0.026217      0.037235      0.050185   
50%        0.119635      0.148590      0.181905      0.219260      0.270145   
75%        0.510915      0.607150      0.698085      0.781490      0.835672   
max        1.020000      1.019970      1.019970      1.019940      1.019990   

       ...       900_dst       910_dst       920_dst       930_dst  \
count  ...  8.001000e+03  7.970000e+03  7.962000e+03  8.022000e+03   
mean   ...  7.976805e-11  6.761949e-11  6.687656e-11  8.491835e-11   
std    ...  3.778529e-10  3.336177e-10  3.453981e-10  4.335765e-10   
min    ...  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    ...  3.501349e-19  1.019189e-19  7.604346e-20  3.993014e-19   
50%    ...  1.067215e-15  6.275528e-16  4.799711e-16  1.149284e-15   
75%    ...  1.168691e-12  9.884604e-13  8.891546e-13  1.224508e-12   
max    ...  1.167515e-08  1.215034e-08  1.288854e-08  1.400634e-08   

            940_dst       950_dst       960_dst       970_dst       980_dst  \
count  7.982000e+03  7.987000e+03  7.992000e+03  7.931000e+03  8.081000e+03   
mean   7.511190e-11  4.270426e-11  2.498465e-11  4.047717e-11  1.153774e-10   
std    3.831916e-10  2.293432e-10  1.639150e-10  2.134495e-10  5.179299e-10   
min    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    1.574964e-19  0.000000e+00  0.000000e+00  0.000000e+00  5.669218e-18   
50%    6.566552e-16  8.391564e-17  2.525680e-17  1.231848e-16  3.395726e-15   
75%    9.503366e-13  3.186890e-13  1.234955e-13  2.958971e-13  2.392410e-12   
max    1.343132e-08  6.461523e-09  5.773867e-09  5.529940e-09  1.013262e-08   

            990_dst  
count  8.013000e+03  
mean   2.364072e-10  
std    1.058272e-09  
min    0.000000e+00  
25%    6.574627e-17  
50%    1.717923e-14  
75%    6.942183e-12  
max    2.682289e-08  

[8 rows x 71 columns]

In [8]:
label = data[['hhb','hbo2','ca','na']]

data.drop(columns=['id','hhb','hbo2','ca','na'], inplace=True)
test.drop(columns=['id'], inplace=True)

In [12]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 71 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   rho      10000 non-null  int64  
 1   650_src  10000 non-null  float64
 2   660_src  10000 non-null  float64
 3   670_src  10000 non-null  float64
 4   680_src  10000 non-null  float64
 5   690_src  10000 non-null  float64
 6   700_src  10000 non-null  float64
 7   710_src  10000 non-null  float64
 8   720_src  10000 non-null  float64
 9   730_src  10000 non-null  float64
 10  740_src  10000 non-null  float64
 11  750_src  10000 non-null  float64
 12  760_src  10000 non-null  float64
 13  770_src  10000 non-null  float64
 14  780_src  10000 non-null  float64
 15  790_src  10000 non-null  float64
 16  800_src  10000 non-null  float64
 17  810_src  10000 non-null  float64
 18  820_src  10000 non-null  float64
 19  830_src  10000 non-null  float64
 20  840_src  10000 non-null  float64
 21  850_src  1000

In [13]:
cat_columns = [] 
num_columns = [c for c in data.columns if c not in cat_columns]

In [ ]:
cat_columns, num_columns

In [17]:
np.mean(data)

rho        1.756800e+01
650_src    1.802125e-01
660_src    2.035285e-01
670_src    2.298040e-01
680_src    2.591581e-01
               ...     
950_dst    4.270426e-11
960_dst    2.498465e-11
970_dst    4.047717e-11
980_dst    1.153774e-10
990_dst    2.364072e-10
Length: 71, dtype: float64

In [58]:
def preprocess(x_train, x_valid, x_test):
    
    tmp_x_train = x_train.copy()
    tmp_x_valid = x_valid.copy()
    tmp_x_test  = x_test.copy()
    
    # 결측치 처리 : 보간법 - 데이콘 코드 공유, 실무자에 따르면 스펙트럼 데이터를 단순 평균으로 대체하는 것은 큰 오류를 범할 수 있으므로 보간법으로 대체해야 한다고함
    tmp_x_train = tmp_x_train.interpolate(methods='linear', axis=1)
    tmp_x_valid = tmp_x_valid.interpolate(methods='linear', axis=1)
    tmp_x_test = tmp_x_test.interpolate(methods='linear', axis=1)     
    
    # 스케일링
    x_train_mean = np.mean(tmp_x_train[num_columns], axis=0)
    x_train_std  = np.std(tmp_x_train[num_columns], axis=0)

    tmp_x_train.loc[:, num_columns] = (tmp_x_train[num_columns] - x_train_mean) / (x_train_std + 1e-4)
    tmp_x_valid.loc[:, num_columns] = (tmp_x_valid[num_columns] - x_train_mean) / (x_train_std + 1e-4)
    tmp_x_test.loc[:, num_columns] = (tmp_x_test[num_columns] - x_train_mean) / (x_train_std + 1e-4)
    
    # 인코딩 : 전부 연속형 변수라 불필요
    
    # 리턴
    return tmp_x_train, tmp_x_valid, tmp_x_test

In [ ]:
tmp_x_train = x_train.copy()
tmp_x_valid = x_valid.copy()
tmp_x_test  = x_test.copy()

In [19]:
from sklearn.model_selection import KFold # KFold일지 StratifiedKFold를 사용해야하는지 고민해보세요.
kf = KFold(n_splits=5, shuffle=True, random_state=2020)

In [23]:
from sklearn.linear_model import Lasso
from sklearn.linear_model import Ridge
from sklearn.ensemble import RandomForestRegressor
from lightgbm import LGBMRegressor

from sklearn.multioutput import MultiOutputRegressor
from sklearn.metrics import mean_absolute_error

In [59]:
#######Lasso 학습

val_scores1 = list()

for i, (trn_idx, val_idx) in enumerate(kf.split(data, label)) :
    x_train, y_train = data.iloc[trn_idx, :], label.iloc[trn_idx,:]
    x_valid, y_valid = data.iloc[val_idx, :], label.iloc[val_idx,:]
    
    # 전처리
    x_train, x_valid, _ = preprocess(x_train, x_valid, test) # _ : 함수에서 값은 나오지만 처리하지는 않겠다는 의미
    
    # 모델 정의
    ls = MultiOutputRegressor(Lasso(random_state=2020))
    
    # 모델 학습
    ls.fit(x_train, y_train)

    # 훈련, 검증 데이터 
    trn_mae = mean_absolute_error(y_train, ls.predict(x_train))
    val_mae = mean_absolute_error(y_valid, ls.predict(x_valid))
    print('{} Fold, train mae : {:.4f}, validation mae : {:.4f}'.format(i, trn_mae, val_mae))
    
    val_scores1.append(val_mae)

#교차 검증
print('Cross Validation Score : {:.4f}%'.format(np.mean(val_scores1)))

0 Fold, train mae : 1.7690, validation mae : 1.7868
1 Fold, train mae : 1.7697, validation mae : 1.7843
2 Fold, train mae : 1.7768, validation mae : 1.7564
3 Fold, train mae : 1.7769, validation mae : 1.7555
4 Fold, train mae : 1.7706, validation mae : 1.7807
Cross Validation Score : 1.7727%


In [61]:
#######Ridge 학습

val_scores2 = list()

for i, (trn_idx, val_idx) in enumerate(kf.split(data, label)) :
    x_train, y_train = data.iloc[trn_idx, :], label.iloc[trn_idx,:]
    x_valid, y_valid = data.iloc[val_idx, :], label.iloc[val_idx,:]
    
    # 전처리
    x_train, x_valid, _ = preprocess(x_train, x_valid, test) # _ : 함수에서 값은 나오지만 처리하지는 않겠다는 의미
    
    # 모델 정의
    rg = MultiOutputRegressor(Ridge(random_state=2020))
    
    # 모델 학습
    rg.fit(x_train, y_train)

    # 훈련, 검증 데이터 
    trn_mae = mean_absolute_error(y_train, rg.predict(x_train))
    val_mae = mean_absolute_error(y_valid, rg.predict(x_valid))
    print('{} Fold, train mae : {:.4f}, validation mae : {:.4f}'.format(i, trn_mae, val_mae))
    
    val_scores2.append(val_mae)

#교차 검증
print('Cross Validation Score : {:.4f}%'.format(np.mean(val_scores2)))

0 Fold, train mae : 1.7640, validation mae : 1.7891
1 Fold, train mae : 1.7650, validation mae : 1.7906
2 Fold, train mae : 1.7722, validation mae : 1.7601
3 Fold, train mae : 1.7722, validation mae : 1.7584
4 Fold, train mae : 1.7649, validation mae : 1.7864
Cross Validation Score : 1.7769%


In [62]:
#######Randomforest 학습

val_scores3 = list()

for i, (trn_idx, val_idx) in enumerate(kf.split(data, label)) :
    x_train, y_train = data.iloc[trn_idx, :], label.iloc[trn_idx,:]
    x_valid, y_valid = data.iloc[val_idx, :], label.iloc[val_idx,:]
    
    # 전처리
    x_train, x_valid, _ = preprocess(x_train, x_valid, test) # _ : 함수에서 값은 나오지만 처리하지는 않겠다는 의미
    
    # 모델 정의
    rf = MultiOutputRegressor(RandomForestRegressor(n_estimators=10, max_depth=50, random_state=2020, n_jobs = 6))
    
    # 모델 학습
    rf.fit(x_train, y_train)

    # 훈련, 검증 데이터 
    trn_mae = mean_absolute_error(y_train, rf.predict(x_train))
    val_mae = mean_absolute_error(y_valid, rf.predict(x_valid))
    print('{} Fold, train mae : {:.4f}, validation mae : {:.4f}'.format(i, trn_mae, val_mae))
    
    val_scores3.append(val_mae)

#교차 검증
print('Cross Validation Score : {:.4f}%'.format(np.mean(val_scores3)))

0 Fold, train mae : 0.6956, validation mae : 1.7321
1 Fold, train mae : 0.6988, validation mae : 1.7030
2 Fold, train mae : 0.6915, validation mae : 1.6985
3 Fold, train mae : 0.6869, validation mae : 1.7064
4 Fold, train mae : 0.6896, validation mae : 1.7272
Cross Validation Score : 1.7134%


In [63]:
#######LGB 학습

val_scores4 = list()

for i, (trn_idx, val_idx) in enumerate(kf.split(data, label)) :
    x_train, y_train = data.iloc[trn_idx, :], label.iloc[trn_idx,:]
    x_valid, y_valid = data.iloc[val_idx, :], label.iloc[val_idx,:]
    
    # 전처리
    x_train, x_valid, _ = preprocess(x_train, x_valid, test) # _ : 함수에서 값은 나오지만 처리하지는 않겠다는 의미
    
    # 모델 정의
    lgb = MultiOutputRegressor(LGBMRegressor(n_estimators=150, max_depth=10, random_state=2020, n_jobs = 6))
    
    # 모델 학습
    lgb.fit(x_train, y_train)

    # 훈련, 검증 데이터 
    trn_mae = mean_absolute_error(y_train, lgb.predict(x_train))
    val_mae = mean_absolute_error(y_valid, lgb.predict(x_valid))
    print('{} Fold, train mae : {:.4f}, validation mae : {:.4f}'.format(i, trn_mae, val_mae))
    
    val_scores4.append(val_mae)

#교차 검증
print('Cross Validation Score : {:.4f}%'.format(np.mean(val_scores4)))

0 Fold, train mae : 0.7594, validation mae : 1.2093
1 Fold, train mae : 0.7665, validation mae : 1.2090
2 Fold, train mae : 0.7679, validation mae : 1.1813
3 Fold, train mae : 0.7615, validation mae : 1.1965
4 Fold, train mae : 0.7615, validation mae : 1.2051
Cross Validation Score : 1.2002%


In [68]:
####### oof 앙상블

val_scores_fnl = list()
oof_pred = np.zeros((data.shape[0], label.shape[1]))

for i, (trn_idx, val_idx) in enumerate(kf.split(data, label)):
    x_train, y_train = data.iloc[trn_idx, :], label.iloc[trn_idx,:]
    x_valid, y_valid = data.iloc[val_idx, :], label.iloc[val_idx,:]
    
    # 전처리
    x_train, x_valid, x_test = preprocess(x_train, x_valid, test)
    
    # 모델 정의
    lgb = MultiOutputRegressor(LGBMRegressor(n_estimators=150, max_depth=10, random_state=2020, n_jobs = 6))
    
    # 모델 학습
    lgb.fit(x_train, y_train)

    # 훈련, 검증 데이터 MAE 확인
    trn_mae = mean_absolute_error(y_train, lgb.predict(x_train))
    val_mae = mean_absolute_error(y_valid, lgb.predict(x_valid))
    print('{} Fold, train MAE : {:.4f}4, validation MAE : {:.4f}'.format(i, trn_mae, val_mae))
    
    val_scores_fnl.append(val_mae)
    
    # MAE니까 Fold 전체의 평균
    oof_pred += lgb.predict(x_test) / 5
    

# 교차 검증 
print('Cross Validation Score : {:.4f}'.format(np.mean(val_scores_fnl)))


0 Fold, train MAE : 0.75944, validation MAE : 1.2093
1 Fold, train MAE : 0.76654, validation MAE : 1.2090
2 Fold, train MAE : 0.76794, validation MAE : 1.1813
3 Fold, train MAE : 0.76154, validation MAE : 1.1965
4 Fold, train MAE : 0.76154, validation MAE : 1.2051
Cross Validation Score : 1.2002


######### 앙상블 결과로 cv값이 1.2가 나왔는데, 제출 후 점수는 1.18로 약간 상승했습니다

In [74]:
tmp = pd.DataFrame(oof_pred, columns=['hhb','hbo2','ca','na'])
tmp.head()

hhb      hbo2         ca        na
0   8.832935  4.347779   9.330061  3.968840
1   6.075171  3.710824   8.151994  2.482501
2  10.269393  4.834989  10.574285  4.046433
3   8.593961  4.062690  10.700909  3.566748
4   6.539964  3.585697   8.797350  3.950380

In [75]:
submit_path = join('dacon5', 'sample_submission.csv')
submit = pd.read_csv(submit_path)

In [76]:
y_test_layer1 = tmp.iloc[:,0]
y_test_layer2 = tmp.iloc[:,1]
y_test_layer3 = tmp.iloc[:,2]
y_test_layer4 =tmp.iloc[:,3]

In [77]:
submit['hhb'] = y_test_layer1
submit['hbo2'] = y_test_layer2
submit['ca'] = y_test_layer3
submit['na'] = y_test_layer4


submit_file_name = 'dacon5.csv'
submit.to_csv(submit_file_name, index=False)